homeassistant/components/roborock/__init__.py

```diff
@@ -24,27 +31,37 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up roborock from a config entry."""
     _LOGGER.debug("Integration async setup entry: %s", entry.as_dict())
 
-    user_data = UserData(entry.data[CONF_USER_DATA])
-    api_client = RoborockClient(entry.data[CONF_USERNAME], entry.data[CONF_BASE_URL])
+    user_data = UserData.from_dict(entry.data[CONF_USER_DATA])
+    api_client = RoborockApiClient(entry.data[CONF_USERNAME], entry.data[CONF_BASE_URL])
     _LOGGER.debug("Getting home data")
     home_data = await api_client.get_home_data(user_data)
     _LOGGER.debug("Got home data %s", home_data)
 
-    device_map: dict[str, RoborockDeviceInfo] = {}
-    devices = home_data.devices + home_data.received_devices
+    devices_info: dict[str, RoborockDeviceInfo] = {}
```

Comment:
There is a typing issue here.  This is `dict[str, RoborockDeviceInfo]` but the coordinator expects `dict[str, RoborockHassDeviceInfo]` i think.

Plus, creating a dict then either deleting entries or replacing entries with a different type is harder to follow. I'd say just iterate over the devices and populate the final dict of  `RoborockHassDeviceInfo`.

Make products into a dict as well to simplify and shorten the loop

Reply:
I think I need three separate dicts, especially after switching to composition.

Mqtt Client takes a dict of device info

Local Client takes a dict of local device info

Then the coordinator takes a dict of hass device info

Is there a better way that I am missing?

Reply:
I don't see anything super obvious you're missing, but here are some ideas:
- inline some of the dict creation in the client library constructors
- push the client libraries creation all into the update coordinator (e.g. the coordinator just takes a list of devices as input, and hides all the client library details from init. doesn't necessarily solve the issue, but you may find that this gets easier to manage in one place).   for example, do you still want to create the entities anyway and just mark them `unavailable` rather than not creating them at all?
- simplify the client library constructors e.g. the client library itself takes a dict, but it could just also take a list of devices
- generally simplify the client library to abstract all of this away to provide the APIs you actually want to use

Reply:
Moved it into constructor and inlined some of the dict creation, could potentially inline inside the coordinator when I create the local_devices_info and hass_devices_info, but because they need networking, I wasn't sure if it was efficient, as they would both end up reading from the networking dict as well.